### PR TITLE
values need to be bytes when writing back to ldap

### PR DESCRIPTION
**Pull request by angeloudy**
### What issues does this PR fix or reference?
fix for #48599

### CODE REVIEW

File: salt/modules/ldap3.py [hunk 1]
```diff
@@ -572,19 +572,16 @@ def change(connect_spec, dn, before, after):
     # convert the "iterable of values" to lists in case that's what
     # modifyModlist() expects (also to ensure that the caller's dicts
     # are not modified)
-    before = dict(((attr, list(vals))
+    before = dict(((attr, salt.utils.data.encode(vals))
                    for attr, vals in six.iteritems(before)))
-    after = dict(((attr, list(vals))
+    after = dict(((attr, salt.utils.data.encode(vals))
```
Comment (terminalmage) [inline salt/modules/ldap3.py]: I'm not familiar with this module, but `list()` has been there from the beginning (according to git log). Can you explain why you removed it here?

**Comment by dwoz** [comment]
Should the encoding/decoding be specific to the ldap server? It looks like this PR is basing the encoding/decoding based on the client's system settings. Have you thought about what happens if the client system is using a non utf-8 encoding?

File: salt/modules/ldap3.py (hunk 1 above)
Comment (angeloudy) [reply salt/modules/ldap3.py]: In fact, it would be safe to use `salt.utils.data.encode(list(vals))` because vals could be type `OrderedSet`

**Comment by angeloudy** [comment]
@dwoz 
That's a good point!
I have no idea how to get the encoding.
What do you suggest?

**Comment by dwoz** [comment]
@angeloudy I think it's best to use utf-8 here (it looks like that is what LDAPv3 expects by default). We can make it configurable down the road if needed

**Comment by cachedout** [comment]
@dwoz and @angeloudy There is also `__salt_system_encoding__` that should be available.

**Comment by angeloudy** [comment]
@cachedout should `salt.utils.data.encode` use `__salt_system_encoding__` by default?
I did have any 'utf8' hardcoded in my changes.

**Comment by terminalmage** [comment]
@angeloudy It would try utf8 first, and then try the detected encoding if that raises a `UnicodeEncodeError`.

**Comment by rallytime** [comment]
@angeloudy Are you planning on making the requested changes here?

**Comment by angeloudy** [comment]
@rallytime what requested changes.
I am fine with my current version.

**Comment by angeloudy** [comment]
@terminalmage @rallytime 
if an `UnicodeEncodeError` exception catching is necessary, shouldn't it be done inside `salt.utils.data.encode`?

**Comment by rallytime** [comment]
@terminalmage and @dwoz can you comment/re-review?

**Comment by angeloudy** [comment]
Hi, any plans to merge this pull request so that I don't have to maintain my own patch?
I don't understand why are you guys worrying about UnicodeEncodeError, the encode method is being used everywhere in saltstack.

**Review by dwoz — APPROVED** [review]

File: salt/modules/ldap3.py (hunk 1 above)
Comment (terminalmage) [reply salt/modules/ldap3.py]: Then you should probably make that change.

**Review by terminalmage — CHANGES_REQUESTED** [review]
I don't think anyone is worrying about an encoding error. You still haven't made the suggested changes from a few months ago, which is probably why this hasn't been merged.

**Review by terminalmage — APPROVED** [review]

**Pull request merged**